File: segmentation/src/bottomUpSegmentation.cpp
/**
 *
 * @file	bottomUpSegmentation.cpp
 * @brief   Bottom-UP segmentation approach,
 *          based on a Non Linear Regression Approximation algorithm
 * @author  Henri NG
 * @version 1.0
 * @date	20/05/2014
 *
 */

#include "polyfit.hpp"
#include "segmentation.h"

Segment bottomUpSegmentation(const vector<float>& data, int nb_seg)
{

    // Initialize variables
	int data_size = data.size();
	Segment seg(data_size);

    // Harmonize data
	vector<float> datah = harmonize(data);
	float stdev = standard_deviation(datah);
    for (int i = 0; i < data_size; i++)
        datah[i] /= stdev;
	vector<float> curReconstruct = datah;

	// Initialize the merge cost of the segments
	// in the "fine segmented representation"
	for (int i = 0; i < seg.nbSeg - 1; i++) {
		seg.best[i] = compute_best(datah, seg, i, 1);
		seg.mc[i]   = compute_mc(datah, seg, i, 1);
	}	

	// Keep merging the lowest cost segments until only 'nb_seg' remain.
	int vind;
	float vmin;
	float err_recons = 0;

	while ((err_recons < 0.006) || (seg.size() > nb_seg)) {

        // Find the location "vind", of the cheapest merge.
		min_vector(seg.mc, &vmin, &vind);

		// The typical case, neither of the two segments
		// to be merged are end segments
		if ((vind > 0) && (vind < seg.nbSeg - 2)) {
			//std::cout << "Case 1 : " << vind << std::endl;
			//std::cout << "nb_seg : " << seg.nbSeg << std::endl;
			compute_recons(curReconstruct, seg, vind);
			seg.best[vind]   = compute_best(datah, seg, vind, 2);
			seg.mc[vind]     = compute_mc(datah, seg, vind, 2);
			seg.rightx[vind] = seg.rightx[vind + 1];
			seg.removePoint(vind + 1);
			vind--;
			seg.best[vind]   = compute_best(datah, seg, vind, 1);
			seg.mc[vind]     = compute_mc(datah, seg, vind, 1);
		}

		// Special case: The leftmost segment must be merged.
		else if (vind == 0) {
			//std::cout << "Case 2 : " << vind << std::endl;
			//std::cout << "nb_seg : " << seg.nbSeg << std::endl;
			compute_recons(curReconstruct, seg, vind);
			seg.best[vind]   = compute_best(datah, seg, vind, 2);
			seg.mc[vind]     = compute_mc(datah, seg, vind, 2);
			seg.rightx[vind] = seg.rightx[vind + 1];
			seg.removePoint(vind + 1);
		}

		// Special case: The rightmost segment must be merged.
		else {
			//std::cout << "Case 3 : " << vind << std::endl;
			//std::cout << "nb_seg : " << seg.nbSeg << std::endl;
			compute_recons(curReconstruct, seg, vind);
			seg.rightx[vind] = seg.rightx[vind + 1];
			seg.mc[vind]     = FLT_MAX;
			seg.removePoint(vind + 1);
			vind--;
			seg.best[vind]   = compute_best(datah, seg, vind, 1);
			seg.mc[vind]     = compute_mc(datah, seg, vind, 1);
		}

	err_recons = compute_error(datah, curReconstruct);
	//std::cout << "err_recons = " << err_recons << std::endl;

	}

    vector<float> best;
	for (int i = 0; i < seg.nbSeg; i++) {
        seg.best[i] = compute_best(datah, seg, i, 0);
        best = seg.best[i];
        seg.lefty[i]  = best[0];
        seg.righty[i] = best[best.size() - 1];
        seg.mc[i] = compute_mc(datah, seg, i, 0);
	}

    seg.resize();

	return seg;

}

float compute_mc(const vector<float>& data,
                 const Segment& seg,
                 int ind, int delay)
{
    int left  = seg.leftx[ind];
	int right = seg.rightx[ind + delay];
	float tmp   = 0;
	float accum = 0;
	vector<float> best = seg.best[ind];
	//std::cout << "best = ";
    //print_vector(best);
	for (int k = left; k < right; k++) {
		//std::cout << data[k] << "\t" << best[k - left] << std::endl;
		tmp = data[k] - best[k - left];
		accum += (tmp * tmp);
	}
	return accum;
}

vector<float> compute_best(const vector<float>& data,
						   const Segment& seg,
						   int ind, int delay)
{
	float left  = seg.leftx[ind];
	float right = seg.rightx[ind + delay];
	float gap   = right - left + 1;
	float deg   = 1;
	vector<float> best(gap, 0);
	vector<float> x = step_vector(left, right, deg);
	vector<float> y(data.begin() + left, data.begin() + right + 1);
	vector<float> coef = mathalgo::polyfit(x, y, 1);
	//std::cout << "x = ";
	//print_vector(x);
	//std::cout << "y = ";
	//print_vector(y);
	std::reverse(coef.begin(), coef.end());
	//std::cout << "c = ";
	//print_vector(coef);
	for (int i = 0; i < gap; i++)
		best[i] = coef[0] * x[i] + coef[1];
	//std::cout << "b = ";
	//print_vector(best);
	return best;
}

void compute_recons(vector<float>& data, const Segment& seg, int ind)
{
	int left  = seg.leftx[ind];
	int right = seg.rightx[ind + 1];
	vector<float> best = seg.best[ind];
	for (int k = left; k <= right; k++)
		data[k] = best[k - left];
}

float compute_error(const vector<float>& data, const vector<float>& recd)
{
	int data_size = data.size();
	float err_recons = 0;
	float accum = 0;
	float tmp   = 0;
	for (int k = 0; k < data_size; k++) {
		//std::cout << data[k] << "\t" << recd[k] << std::endl;
		tmp = data[k] - recd[k];
		//std::cout <<  tmp << std::endl;
		accum += (tmp * tmp);
	}
	err_recons = sqrtf(accum) / data_size;
	return err_recons;
}

File: segmentation/matlab/bottomUpSegmentation.m
function segment = bottomUpSegmentation(data,num_segments,forceplot)

% The basic algorithm works by creating a fine segmented representation then merging the lowest cost segments until only 'num_segments' remain.

left_x       = 1:2:(size(data,1) - 1);    % Find the left x values vector for the "fine segmented representation".
right_x      = left_x + 1;                  % Find the right x values vector for the "fine segmented representation".
right_x(end) = size(data,1);                % Special case, the rightmost endpoint.
number_of_segments = length(left_x );       % Calculate the number of segments in the initial "fine segmented representation".

data = (data - mean(data));
data = data ./ std(data);
curReconstruct = data;

segments = [];
if nargin > 2
    figure;
    plot(data);
end

% Initialize the segments in the "fine segmented representation". 

for i = 1 : number_of_segments 
   
   segment(i).lx = left_x(i);
   segment(i).rx = right_x(i);
   segment(i).mc = inf;
   segment(i).best = [];
end;

% Initialize the merge cost of the segments in the "fine segmented representation". 

for i = 1 : number_of_segments - 1
   coef = polyfit([segment(i).lx :segment(i+1).rx]',data(segment(i).lx :segment(i+1).rx),1);
   segment(i).best = (coef(1)*( [segment(i).lx :segment(i+1).rx]' ))+coef(2);
   segment(i).mc = sum((data([segment(i).lx :segment(i+1).rx]')-segment(i).best).^2);
end;

% Keep merging the lowest cost segments until only 'num_segments' remain. 

errRecon = 0;

while errRecon < 0.006 || length(segment) > num_segments
   
   
   [value, i ] = min([segment(:).mc]);                              % Find the location "i", of the cheapest merge.      
   
   if i > 1 && i < length(segment) - 1								% The typical case, neither of the two segments to be merged are end segments
      
       
        curReconstruct(segment(i).lx:segment(i+1).rx) = segment(i).best;
    	coef = polyfit([segment(i).lx :segment(i+2).rx]',data(segment(i).lx :segment(i+2).rx),1);
    	segment(i).best = (coef(1)*( [segment(i).lx :segment(i+2).rx]' ))+coef(2);
       
        segment(i).mc = sum((data([segment(i).lx :segment(i+2).rx]')-segment(i).best).^2);

	 	segment(i).rx = segment(i+1).rx;
    	segment(i+1) = [];
    
    	i = i - 1;
    
        coef = polyfit([segment(i).lx :segment(i+1).rx]',data(segment(i).lx :segment(i+1).rx),1);
    	segment(i).best = (coef(1)*( [segment(i).lx :segment(i+1).rx]' ))+coef(2);
    
        segment(i).mc = sum((data([segment(i).lx :segment(i+1).rx]')-segment(i).best).^2);
       
   elseif i == 1                                                    % Special case: The leftmost segment must be merged.
       
        curReconstruct(segment(i).lx:segment(i+1).rx) = segment(i).best;
	    coef = polyfit([segment(i).lx :segment(i+2).rx]',data(segment(i).lx :segment(i+2).rx),1);
    	segment(i).best = (coef(1)*( [segment(i).lx :segment(i+2).rx]' ))+coef(2);
    	segment(i).mc = sum((data([segment(i).lx :segment(i+2).rx]')-segment(i).best).^2);

	 	segment(i).rx = segment(i+1).rx;
        segment(i+1) = [];
              
   else                                                             % Special case: The rightmost segment must be merged.
     
       
       curReconstruct(segment(i).lx:segment(i+1).rx) = segment(i).best;
       segment(i).rx = segment(i+1).rx;
       segment(i).mc = inf;
       segment(i+1) = [];
    
       i = i - 1;
       
       coef = polyfit([segment(i).lx :segment(i+1).rx]',data(segment(i).lx :segment(i+1).rx),1);
       segment(i).best = (coef(1)*( [segment(i).lx :segment(i+1).rx]' ))+coef(2);
    
       segment(i).mc = sum((data([segment(i).lx :segment(i+1).rx]')-segment(i).best).^2);

   end;
   
   errRecon = sqrt(sum((data - curReconstruct) .^ 2)) ./ length(data);
           
end;


%----------------------------------------------
residuals=[];

for i = 1 : length(segment) 
   
        coef = polyfit([segment(i).lx :segment(i).rx]',data(segment(i).lx :segment(i).rx),1);
    	best = (coef(1)*( [segment(i).lx :segment(i).rx]' ))+coef(2);
        residuals = [ residuals ; sum((data([segment(i).lx :segment(i).rx]')-best).^2)];
end;




%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%% If the user passed in an extra argument, then plot the original time series, together with the segments %%%%%%%%%%%%%


if nargin > 2
    figure;
    hold on;
    %plot(data,'r'); 
end; 

   
    temp = [];
    for i = 1 : length(segment) 
   
      coef = polyfit([segment(i).lx :segment(i).rx]',data(segment(i).lx :segment(i).rx),1);
      best = (coef(1)*( [segment(i).lx :segment(i).rx]' ))+coef(2);
      
      segment(i).ly = best(1);
      segment(i).ry = best(end);
      segment(i).mc = residuals(i);
      
      if nargin > 2
            plot([segment(i).lx:segment(i).rx]', best,'b');
      end;
  
      temp = [temp; [segment(i).ly segment(i).ry]];
   
    end;

    
if nargin > 2
    for i = 1 : length(segment)  - 1 
        plot([segment(i).rx :segment(i+1).lx]', [ temp(i,2) temp(i+1,1)  ],'g');
    end;
end;
